ql/src/gen/vectorization/ExpressionTemplates/ColumnArithmeticColumn.txt

```diff
@@ -34,20 +34,17 @@ public class <ClassName> extends VectorExpression {
 
   private static final long serialVersionUID = 1L;
 
-  private final int colNum1;
   private final int colNum2;
```

Comment:
It will be better to have the two input columns together in the same place. But it is a good idea to move this to parent class and share the assigning code for all subclasses. Can we introduce two subclasses VectorUnaryExpression and VectorBinaryExpression( or similar name) for VectorExpression and then extend other classes from either one of these classes.

Reply:
I agree that the current solution is not really clean by having only the first column put into VectorExpression
a couple of notes here, which needs to be discussed before proceeding with this huge refactor (which I'm happy to do once we 100% certain about the "perfect" solution):

1. unary, binary is not enough, unfortunately, we have even expressions involving even more cols, this is not a problem, we have the language support for that :) tertiary, quaternary...

2. what's confusing is, how to show with simple class names that unary/binary/... is only a story about the input columns? an expression can have constants too, e.g. in IfExprScalarScalar.txt:
```
     this.arg1Column = arg1Column;
     this.arg2Scalar = arg2Scalar;
     this.arg3Scalar = arg3Scalar;
```
in our terminology here, this is a unary expression because of arg1Column + scalars, but in reality, it's obviously not a unary function...

3. with subclasses, we'll have to implement a general VectorExpression.setInputColumnNum(int i, int j, int k, ...vararg), otherwise, we won't be able to change the input column numbers (which is important, this was the intention of this huge vector expression refactor), I think this will simply work by simply overriding vararg method in subclasses

Reply:
a) regarding input columns, a simple solution is could be https://github.com/apache/hive/pull/2099/commits/5db4c60aec34df291a1d1d85be319c865adcd1ad
created convenience constructors, subclasses should be refactored accordingly now with care

b) I was also wondering about having an array of inputcolumns as:
```
public int[] inputColumnNums = new int[] { -1, -1, -1 };
```
this could be more general, more easily extendable, and needs a bit more refactoring (every existing occurrence of inputColumn should becode inputColumnNums[0])

so regarding the class naming, we won't change anything (to unary, binary, ...) just use the proper constructor in subclasses

Reply:
what do you think about this @ramesh0201?
I think a general input col array would be nice (option b) )

however, there some rare cases where it's not obvious which position should be used, but it's up to agreement e.g.:
IfExprScalarColumn.txt
```
protected final int arg1Column;
protected final <OperandType2> arg2Scalar;
protected final int arg3Column;
```
this is tricky because there is a scalar interleaved into the columns, input col array might look like:
1. new int[] { arg1Column, -1, arg3Column};
to emphasize that that the second argument is a scalar, so we'll refactor as:
```
arg3Column => inputColumnNums[2]
```

2. new int[] { arg1Column, arg3Column, -1};
to ignore the fact that there is an interleaved scalar input, so we'll refactor as:
```
arg3Column => inputColumnNums[1]
```

Reply:
@abstractdog Option a also looks good to me. If you think that will be better then we can go with that. 

Otherwise if we choose option b we can choose 2) new int[] { arg1Column, arg3Column, -1}; This way it is always easy to know only first n columns are used, rather than confused with the interleaving part

Reply:
okay, I'm trying to proceed with 2) with your advice, because an array is nicer than having fields, not to mention if we're adding convenience methods for processing input columns (simple loop vs. mentioning all the fields)